src/oscar/navigation_menu_registry.py

```diff
@@ -0,0 +1,237 @@
+
+    @property
+    def is_placeholder(self):
+        return self.label == "" and self.url_name == "" and self.icon == ""
+
+    @classmethod
+    def from_dict(cls, _dict):
+        """Should help ease migration process from menus declared through `settings.OSCAR_DASHBOARD_NAVIGATION`"""
+        children = []
+        for index, child_dict in enumerate(_dict.pop("children", [])):
+            menu = cls.from_dict(child_dict)
+            menu.auto_position_if_applicable(index)
+            children.append(menu)
+        menu = cls(**_dict)
+        menu._children += children
+        return menu
+
+    @property
+    def children(self):
+        return sorted(self._children)
+
+    def add_child(self, menu):
+        """
+        Add child to `self` with the extra advantage of:
+         1. preserving the order of addition.
+         2. merging `menu` to existing menu (with same identity).
+        """
+        for index, child in enumerate(self._children.copy()):
+            if child == menu:
+                self._children[index] = child + menu
+                break
+        else:
+            menu.auto_position_if_applicable(len(self._children))
+            self._children.append(menu)
+        return self
+
+    def add_children(self, children):
+        for child in children:
+            self.add_child(child)
+        return self
+
+    def remove_child(self, identifier):
+        try:
+            self._children.remove(self.placeholder(identifier))
+        except ValueError:
+            raise ValueError(f"Child menu with ID '{identifier}' not found for parent with ID '{self.identifier}'")
+        return self
+
+    def as_navigation_dict(self):
+        return {
+            "label": self.label,
+            "url_name": self.url_name,
+            "icon": self.icon,
+            "children": [menu.as_navigation_dict() for menu in self.children],
+            **self._use_as_provided,
+        }
+
+
+_dashboard_navigation_menu = defaultdict(Menu)
+_marked_for_deregistration = set()
+
+
+def get_legacy_parent_menu(lookup_label_or_index):
+    """Should help ease migration process from menus declared through `settings.OSCAR_DASHBOARD_NAVIGATION`"""
+    is_lookup_by_position = isinstance(lookup_label_or_index, int)
+    if is_lookup_by_position:
+        return settings.OSCAR_DASHBOARD_NAVIGATION[lookup_label_or_index]
+
+    # parent menu is to be identified by label
+    for index, parent_menu in enumerate(settings.OSCAR_DASHBOARD_NAVIGATION):
+        if parent_menu["label"] == lookup_label_or_index:
+            return parent_menu
+    raise ValueError(
+        f"Parent with {'position' if is_lookup_by_position else 'label'} '{lookup_label_or_index}' not found"
+    )
+
+
+def register_dashboard_menu(menu_or_function=None, parent_id=None):
+    if menu_or_function is None:
+        def decorator(decorated_function):
+            register_dashboard_menu(decorated_function, parent_id)
+            return decorated_function
+
```

Comment:
What is the purpose of the above code?

Reply:
The code aims to support conditional menu registration (using annotations) e.g.
```python
@register_dashboard_menu()
def register_exchange_rate_menu():
    if len(get_supported_currencies_setting().keys()) > 1:
        return Menu(label="Exchange rates", icon="fas fa-money-bill-alt", url_name="dashboard:exchange-rate-list")
```